### PR TITLE
Fixes auto scroll down when new messages arrive

### PR DESCRIPTION
**Pull request by Spetastian**

### CODE REVIEW

**Comment by elimisteve** [comment]
@Moe-Shoman This works fine for me, but it was mostly on your machine where we saw this happening.  Can you run this branch on your machine to ensure that this fixes the issue?  Here's how:

```
git remote add spetastian git@github.com:Spetastian/leapchat.git
git fetch --all
git checkout spetastian/message-auto-scroll
```

then `npm run dev` as usual.

**Comment by elimisteve** [comment]
@Spetastian LGTM thus far :-)

**Comment by elimisteve** [comment]
@Spetastian LGTM thus far :-)

File: src/epics/helpers/ChatHandler.js [hunk 1]
```diff
@@ -54,6 +55,7 @@ class ChatHandler {
           .map(JSON.parse)
           .do(contents => {
             this.wsMessageSubject.next({
+              id: guid.create(),
```
Comment (Spetastian) [inline src/epics/helpers/ChatHandler.js]: @elimisteve FYI. In order to easier distinguish messages when deciding wether to scroll down or not i just create a unique ID locally until we have a better solution

File: src/components/chat/MessageBox.js [hunk 2]
```diff
@@ -14,66 +14,15 @@ class MessageBox extends Component {
...
-  hasNewMessages(prevProps){
-    return prevProps.messages.length !== this.props.messages.length;
+  scrollToBottom = () => {
+    playNotification();
+    this.messagesEnd && this.messagesEnd.scrollIntoView({behavior: "smooth"});
   }
 
-  scrollToBottom(){
-    this.messagesEnd.scrollIntoView({behavior: "smooth"});
+  // Separate function to set reference because of https://reactjs.org/docs/refs-and-the-dom.html#caveats
```
Comment (elimisteve) [inline src/components/chat/MessageBox.js]: @Spetastian Interesting!...

File: src/components/chat/MessageBox.js [hunk 3]
```diff
@@ -14,66 +14,15 @@ class MessageBox extends Component {
...
-    this.setState({
-      notifiedIds: notifiedIds.concat(newMessageIds)
-    });
-  }
 
-  hasNewMessages(prevProps){
-    return prevProps.messages.length !== this.props.messages.length;
+  scrollToBottom = () => {
+    playNotification();
+    this.messagesEnd && this.messagesEnd.scrollIntoView({behavior: "smooth"});
```
Comment (Moe-Shoman) [inline src/components/chat/MessageBox.js]: @Spetastian if you remove `{behavior: "smooth"}` it automatically starts at the bottom on page load.

Comment (elimisteve) [reply src/components/chat/MessageBox.js]: @Spetastian Want to make this tweak, then I'll merge?

File: src/components/chat/MessageList.js [hunk 4]
```diff
@@ -5,8 +5,26 @@ import Message from './Message';
 
 class MessageList extends Component {
 
+  shouldComponentUpdate(nextProps, nextState) {
+    return nextProps.messages.length > 0 
+      && nextProps.messages
+        .map(m => m.id)
+        .concat(this.props.messages.map(m => m.id))
+        .reduce((acc, id) => {
+          if(!acc.indexOf(id) === -1){
```
Comment (elimisteve) [inline src/components/chat/MessageList.js]: @Spetastian Is this the same as `acc.indexOf(id) !== -1`?

File: src/components/chat/MessageList.js [hunk 5]
```diff
@@ -5,8 +5,26 @@ import Message from './Message';
...
+            acc.push(id)
+          }
+          return acc;
+        }, [])
+        .length !== this.props.messages;
```
Comment (elimisteve) [inline src/components/chat/MessageList.js]: @Spetastian :+1: I see what you did here... If the number of old messages + the number of new messages !== the number of messages already shown, re-render.

Users can't delete messages right now, which simplifies things.  Later we'll remove out-of-sight messages from the DOM for performance reasons, at which point maybe we should compare the IDs of the newest incoming message and the latest newest rendered message to determine whether a re-render should occur?

Comment (Spetastian) [reply src/components/chat/MessageList.js]: Yeah, somehow we should compare the the subset of the messages that are visible to the user with the "same" subset from the new props.

But this logic is just to determine if we should scroll down. Later we may also have a property that tells the time when there were new messages and instead we compare that prop to determine if we should scroll down. Or show a button that says "new messages, scroll down" or whatever.

**Pull request merged**